src/event.h

```diff
@@ -10,64 +10,80 @@
+#include "base/AsyncCall.h"
 #include "base/Packable.h"
 #include "mem/forward.h"
 
 /* event scheduling facilities - run a callback after a given time period. */
 
 typedef void EVH(void *);
 
-void eventAdd(const char *name, EVH * func, void *arg, double when, int, bool cbdata=true);
-void eventAddIsh(const char *name, EVH * func, void *arg, double delta_ish, int);
-void eventDelete(EVH * func, void *arg);
-void eventInit(void);
-void eventFreeMemory(void);
-int eventFind(EVH *, void *);
-
+/// an entry in the queue of waiting events
 class ev_entry
 {
     MEMPROXY_CLASS(ev_entry);
 
 public:
-    ev_entry(char const * name, EVH * func, void *arg, double when, int weight, bool cbdata=true);
-    ~ev_entry();
-    const char *name;
-    EVH *func;
-    void *arg;
-    double when;
+    ev_entry(char const * name, double when, int weight, const AsyncCall::Pointer &);
 
-    int weight;
-    bool cbdata;
+public:
+    const char *name = nullptr;
+    double when = 0.1;
 
-    ev_entry *next;
+    int weight = 0;
+    AsyncCall::Pointer call;
+
+    ev_entry *next = nullptr;
 };
 
-// manages time-based events
+/// Manages time-based events
 class EventScheduler : public AsyncEngine
 {
 
 public:
     EventScheduler();
     ~EventScheduler();
-    /* cancel a scheduled but not dispatched event */
-    void cancel(EVH * func, void * arg);
-    /* clean up the used memory in the scheduler */
+
+    /// clean up the used memory in the scheduler
     void clean();
-    /* either EVENT_IDLE or milliseconds remaining until the next event */
+
+    /// either EVENT_IDLE or milliseconds remaining until the next event
     int timeRemaining() const;
-    /* cache manager output for the event queue */
+
+    /// cache manager output for the event queue
     void dump(Packable *);
-    /* find a scheduled event */
-    bool find(EVH * func, void * arg);
-    /* schedule a callback function to run in when seconds */
+
+    /// schedule an AsyncCall run in when seconds
+    void schedule(const char *name, const AsyncCall::Pointer &, double when, int weight);
+    /// \deprecated schedule a callback function to run in when seconds. Use AsyncCall instead.
     void schedule(const char *name, EVH * func, void *arg, double when, int weight, bool cbdata=true);
+
+    /// find a scheduled AsyncCall event
+    bool find(const AsyncCall::Pointer &) const;
+    /// \deprecated find a scheduled callback function. Use AsyncCall instead.
+    bool find(EVH * func, void * arg);
+
+    /// \deprecated cancel a scheduled but not dispatched event
+    /// Use AsyncCall API to cancel() the Call instead.
+    void cancel(EVH * func, void * arg);
+
+    /// dial as many due event Calls as possible within timeout seconds.
     int checkEvents(int timeout);
+
     static EventScheduler *GetInstance();
 
 private:
-    static EventScheduler _instance;
     ev_entry * tasks;
 };
 
+/* Legacy C API for event management. Use EventScheduler::GetInstance() API with AsyncCall instead */
```

Comment:
I have been considering adding an Event namespace and src/event/ library for all this code. It will make a few things a lot easier to transition. If you are in agreement I am okay including that change as part of this PR, otherwise will leave as-is.

Reply:
Please do not add an Event namespace. There are not enough public names that we need to group/encapsulate here AFAICT, and Event::Add() reads worse than AddEvent().

A dedicated library for a pair of classes and 400 lines of code feels like a huge overkill to me, especially since a significant portion of that code is legacy code that will be removed in the foreseeable future.

Reply:
bf9b82f now replaces the previously agreed GetInstance() coding style with a Events() singleton fetch that "reads better".

Reply:
To make sense, we ought to make the second half-step: Use `AddEvent()` instead of `Events().AddEvent()` (or whatever the current spelling of those names is).

In the context of this change request, there is no conceptual difference between the old `Foo::GetInstance().DoX()` and the current `Foo().DoX()` half-step. Both suffer from the same set of problems that this change request is targeting! All those problems can and should be addressed by using a `DoX()` public API instead.

----

P.S. I really doubt we have ever agreed that GetInstance() must be used in public APIs like AddEvent(). If we had, it was a mistake.

Reply:
I'm not sure on why you are so insistent on pushing the methods of EventScheduler into the public global scope as a secondary set of wrapper symbols.

If it is just language grammar the equivalent set of sentences is " events: schedule C at when", "events: remove C" etc. that is even how cache.log entries would appear.

Reply:
> I'm not sure on why you are so insistent on pushing the methods of EventScheduler into the public global scope as a secondary set of wrapper symbols.

I would not describe my change request in the above terms, but the primary reasons behind this change request are basic engineering principles such as

* Hide implementation details (e.g., that there is a single scheduler object for all events, and that it has a timeRemaining() method).

* Avoid code duplication (e.g., every event scheduler should not have to repeat the same `Event::GetInstance()->` code snippet).

When followed, these and similar principles reduce bugs, speed up compilation, help with code maintenance/refactoring, etc., etc.